components/profile/my-network-settings/tokens-settings.tsx

```diff
@@ -27,235 +32,152 @@ export default function TokensSettings({
   }
 
   function addTransactionalToken(newToken: Token) {
-    setAllowedTransactionalTokens((oldState) => {
-      if (oldState) return [...oldState, newToken];
-      return [newToken];
-    });
+    setAllowedTransactionalTokensList((oldState) =>  [...(oldState || []), newToken]);
   }
 
-  function changeSelectedTransactionalTokens(newToken: Token[]) {
-    setSelectedTransactionalTokens(newToken);
-  }
+  async function updateTransactionalTokens(tokenType: TokenType){
+    const isTransactional = tokenType === 'transactional';
+    const currentTokens = currentAllowedTokens[tokenType]
 
-  function changeSelectedRewardTokens(newToken: Token[]) {
-    setSelectedRewardTokens(newToken);
-  }
+    const selectedTokens = isTransactional ? selectedTransactionalTokens : selectedRewardTokens;
+    if(!currentTokens || !selectedTokens) return
+    const selectedTokensAddress = selectedTokens.map(({address})=> address)
+    
+    const toAdd = selectedTokensAddress.filter((address)=> !currentTokens.includes(address))
 
-  async function addTransactionalTokens() {
-    const addTransactionalTokens = selectedTransactionalTokens
-      ?.map(({ address }) => {
-        const token = currentAllowedTokens?.transactional?.find((currentAddress) => address === currentAddress);
-        if (!token) return address;
-      })
-      .filter((v) => v);
+    const toRemove = currentTokens.filter((address)=> !selectedTokensAddress.includes(address))
 
-    if (addTransactionalTokens.length > 0) {
-      await state.Service?.active.addAllowedTokens(addTransactionalTokens, true)
-      .then((txInfo) => {
-        processEvent("registry", "changed", state.Service?.network?.lastVisited, {
-          fromBlock: (txInfo as { blockNumber: number }).blockNumber 
-        })
-        getAllowedTokensContract();
-      })
-    }
-  }
+    const transactions = []
 
-  function removeTransactionalTokens() {
-    const removeTransactionalTokens = currentAllowedTokens?.transactional?.map((currentAddress) => {
-      const token = selectedTransactionalTokens?.find(({ address }) => address === currentAddress);
-      if (!token) return currentAddress;
-    }).filter(v => v)
-
-    if (removeTransactionalTokens.length > 0) {
-      state.Service?.active.removeAllowedTokens(removeTransactionalTokens, true)
-      .then((txInfo) => {
-        processEvent("registry", "changed", state.Service?.network?.lastVisited, {
-          fromBlock: (txInfo as { blockNumber: number }).blockNumber 
-        })
-        getAllowedTokensContract();
+    if(toAdd.length) transactions.push(state.Service?.active.addAllowedTokens(toAdd, isTransactional))
+    if(toRemove.length) transactions.push(state.Service?.active.removeAllowedTokens(toRemove, isTransactional))
+
+    Promise.all(transactions).then(([txAdd, txRemove])=>{
+      const blockAdd = (txAdd as { blockNumber: number })?.blockNumber 
+      
+      if(!blockAdd) return;
```

Comment:
Did you tested only removing tokens ? I think that due to this condition it won't work, because if we only remove tokens the `blockAdd` will be `undefined` and the function will always return before the updates.

Reply:
> Did you tested only removing tokens ? I think that due to this condition it won't work, because if we only remove tokens the `blockAdd` will be `undefined` and the function will always return before the updates.

will not, because have a if adding the transactions at array, so if we just remove one token, the action remove will be the first transaction. (line 104 and 105)


` const transactions = []

    if(toAdd.length) transactions.push(state.Service?.active.addAllowedTokens(toAdd, isTransactional))
    if(toRemove.length) transactions.push(state.Service?.active.removeAllowedTokens(toRemove, isTransactional))

    Promise.all(transactions).then(([txAdd, txRemove])=>{
      const blockAdd = (txAdd as { blockNumber: number })?.blockNumber 
      
      if(!blockAdd) return;`

Reply:
As we only need to know the lowest block, maybe we can use a `reduce` on the `Promise.all` result to get it

Reply:
while the explanation is true (since only one index exists at the array, "blockAdd" will in effect be "blockRemove") the naming of the variables will induce in error when re-reading this code later on (hell: I've had a hard time understanding it _now_ and I had this comment explaining xD)

I would _rename_ the `blockAdd` and `blockRemove` to `firstTx` and `secondTx` accordingly, that way when reading the code again the coder will _clearly_ understand what should be happening

Reply:
@moshmage @vhcsilva you guys can check the fix here: https://github.com/bepronetwork/web-network/pull/799